src/test/regress/upgrade/config.py

```diff
@@ -1,68 +1,291 @@
+
+AFTER_CITUS_UPGRADE_COORD_SCHEDULE = "./after_citus_upgrade_coord_schedule"
+BEFORE_CITUS_UPGRADE_COORD_SCHEDULE = "./before_citus_upgrade_coord_schedule"
+MIXED_BEFORE_CITUS_UPGRADE_SCHEDULE = "./mixed_before_citus_upgrade_schedule"
+MIXED_AFTER_CITUS_UPGRADE_SCHEDULE = "./mixed_after_citus_upgrade_schedule"
+
+CITUS_CUSTOM_TEST_DIR = "./tmp_citus_test"
+
+MASTER = "master"
 # This should be updated when citus version changes
-MASTER_VERSION = '10.2'
+MASTER_VERSION = "10.2"
 
 HOME = expanduser("~")
 
 
 CITUS_VERSION_SQL = "SELECT extversion FROM pg_extension WHERE extname = 'citus';"
 
 
-class CitusUpgradeConfig():
+def find_free_port():
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+        s.bind(("", 0))
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        return s.getsockname()[1]
+
 
+class NewInitCaller(type):
+    def __call__(cls, *args, **kwargs):
+        obj = type.__call__(cls, *args, **kwargs)
+        obj.init()
+        return obj
+
+
+class CitusBaseClusterConfig(object, metaclass=NewInitCaller):
     def __init__(self, arguments):
-        self.bindir = arguments['--bindir']
-        self.pre_tar_path = arguments['--citus-pre-tar']
-        self.post_tar_path = arguments['--citus-post-tar']
-        self.pg_srcdir = arguments['--pgxsdir']
-        self.temp_dir = './tmp_citus_upgrade'
-        self.datadir = self.temp_dir + '/data'
+        if "--bindir" in arguments:
+            self.bindir = arguments["--bindir"]
+        self.pg_srcdir = arguments["--pgxsdir"]
+        self.temp_dir = CITUS_CUSTOM_TEST_DIR
+        self.worker_amount = 2
+        self.user = REGULAR_USER_NAME
+        self.is_mx = False
+        self.is_citus = True
+        self.name = type(self).__name__
         self.settings = {
-            'shared_preload_libraries': 'citus',
-            'citus.node_conninfo': 'sslmode=prefer',
-            'citus.enable_version_checks' : 'false'
+            "shared_preload_libraries": "citus",
+            "citus.node_conninfo": "sslmode=prefer",
+            "citus.enable_repartition_joins": True,
+            "client_min_messages": "WARNING",
         }
-        self.mixed_mode = arguments['--mixed']
+        self.new_settings = {}
+        self.add_coordinator_to_metadata = False
+
+    def init(self):
+        self._init_node_name_ports()
+
+        self.datadir = self.temp_dir + "/data"
+        self.datadir += self.name
+        self.input_dir = self.datadir
+        self.output_dir = self.datadir
+        self.output_file = os.path.join(self.datadir, "run.out")
+        if self.worker_amount > 0:
+            self.chosen_random_worker_port = self.random_worker_port()
+        self.settings.update(self.new_settings)
+
+    def coordinator_port(self):
+        return self.node_name_to_ports[COORDINATOR_NAME]
 
-class PGUpgradeConfig():
+    def setup_steps(self):
+        pass
+
+    def random_worker_port(self):
```

Comment:
Instead of random, should we make it round-robin?

Reply:
Why?

Reply:
Because it'd be easier to debug things if anything fails? When I re-run the test suite, I'll get exact same execution.

If I add a new test, the node we run queries would change, but that is expected and normal and even desired.

Reply:
I think random might be good, but we'd need to be able to specify the seed on the command line. And the output of the test should contain the seed.